api4/plugin.go

```diff
@@ -103,52 +104,39 @@ func installPluginFromUrl(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	downloadUrl := r.URL.Query().Get("plugin_download_url")
-
-	if !model.IsValidHttpUrl(downloadUrl) {
-		c.Err = model.NewAppError("installPluginFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
-		return
+	force := false
+	if r.URL.Query().Get("force") == "true" {
+		force = true
 	}
+	downloadUrl := r.URL.Query().Get("plugin_download_url")
 
-	u, err := url.ParseRequestURI(downloadUrl)
-	if err != nil {
-		c.Err = model.NewAppError("installPluginFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
-		return
-	}
+	installFromUrl(c, w, downloadUrl, force)
+}
 
-	if !*c.App.Config().PluginSettings.AllowInsecureDownloadUrl && u.Scheme != "https" {
-		c.Err = model.NewAppError("installPluginFromUrl", "api.plugin.install.insecure_url.app_error", nil, "", http.StatusBadRequest)
+func installMarketplacePlugin(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !*c.App.Config().PluginSettings.Enable {
+		c.Err = model.NewAppError("installMarketplacePlugin", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
 		return
 	}
 
-	client := c.App.HTTPService.MakeClient(true)
-	client.Timeout = INSTALL_PLUGIN_FROM_URL_HTTP_REQUEST_TIMEOUT
-
-	resp, err := client.Get(downloadUrl)
-	if err != nil {
-		c.Err = model.NewAppError("installPluginFromUrl", "api.plugin.install.download_failed.app_error", nil, err.Error(), http.StatusBadRequest)
+	if !*c.App.Config().PluginSettings.EnableMarketplace {
+		c.Err = model.NewAppError("installMarketplacePlugin", "app.plugin.marketplace_disabled.app_error", nil, "", http.StatusNotImplemented)
 		return
 	}
-	defer resp.Body.Close()
-
-	force := false
-	if r.URL.Query().Get("force") == "true" {
-		force = true
-	}
 
-	fileBytes, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		c.Err = model.NewAppError("installPluginFromUrl", "api.plugin.install.reading_stream_failed.app_error", nil, err.Error(), http.StatusBadRequest)
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
 		return
 	}
 
-	manifest, appErr := c.App.InstallPlugin(bytes.NewReader(fileBytes), force)
+	pluginRequest := model.PluginRequestFromReader(r.Body)
+	plugin, appErr := c.App.GetMarketplacePlugin(pluginRequest)
 	if appErr != nil {
 		c.Err = appErr
 		return
 	}
-	w.WriteHeader(http.StatusCreated)
-	w.Write([]byte(manifest.ToJson()))
+
+	installFromUrl(c, w, plugin.DownloadURL, true)
```

Comment:
Currently on the webapp we are not forcing the installing (aka, this is always passed as false). However, all installs at this time are fresh installs given that the functionality to override/upgrade is not exposed because they are being presented with the `Configure` button if it's already installed. Once we implement the upgrade, I believe then we will force a override, so I believe this should be good.